src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttributes<TAttributeValue> : IList<TagHelperAttribute<TAttributeValue>>
+    {
+        private readonly IList<TagHelperAttribute<TAttributeValue>> _attributes;
+
+        public TagHelperAttributes()
+        {
+            _attributes = new List<TagHelperAttribute<TAttributeValue>>();
+        }
+
+        public int Count => _attributes.Count;
+
+        public bool IsReadOnly => false;
+
+        public IEnumerable<string> Keys => _attributes.Select(attribute => attribute.Key);
+
+        public IEnumerable<TAttributeValue> Values => _attributes.Select(attribute => attribute.Value);
+
+        public TagHelperAttribute<TAttributeValue> this[[NotNull] string key] =>
+            _attributes.LastOrDefault(attribute => MatchesAttribute(key, attribute));
+
+        public TagHelperAttribute<TAttributeValue> this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+            set
+            {
+                _attributes[index] = value;
+            }
+        }
+
+        public bool TryGetAttribute([NotNull] string key, out TagHelperAttribute<TAttributeValue> attribute)
+        {
+            attribute = this[key];
+
+            return attribute != null;
+        }
+
+        public bool TryGetAttributes([NotNull] string key, out IEnumerable<TagHelperAttribute<TAttributeValue>> attributes)
+        {
+            attributes = _attributes.Where(attribute => MatchesAttribute(key, attribute));
+
+            return attributes.Any();
+        }
+
+        public void Add([NotNull] string key)
+        {
+            _attributes.Add(new TagHelperAttribute<TAttributeValue>(key)
+            {
+                Minimized = true
+            });
+        }
+
+        public void Add([NotNull] string key, TAttributeValue value)
+        {
+            _attributes.Add(new TagHelperAttribute<TAttributeValue>(key, value));
+        }
```

Comment:
these `Add` helper methods hide too much.  require explicit `TagHelperAttribute` creation in the caller.  if these are just for testing, make them `internal`